src/Agda2Hs/Compile/Data.hs

```diff
@@ -55,17 +61,30 @@ compileData newtyp ds def = do
         DomForall{}     -> genericDocError =<< text "Not supported: indexed datatypes"
       _ -> return ()
 
-compileConstructor :: [Arg Term] -> QName -> C (Hs.QualConDecl ())
+compileConstructor :: [Arg Term] -> QName -> C (Hs.QualConDecl (), Maybe [Hs.Decl ()])
 compileConstructor params c = do
   reportSDoc "agda2hs.data.con" 15 $ text "compileConstructor" <+> prettyTCM c
   reportSDoc "agda2hs.data.con" 20 $ text "  params = " <+> prettyTCM params
   ty <- (`piApplyM` params) . defType =<< getConstInfo c
   reportSDoc "agda2hs.data.con" 20 $ text "  ty = " <+> prettyTCM ty
   TelV tel _ <- telView ty
-  let conName = hsName $ prettyShow $ qnameName c
+
+  let qnameList = qnameToList c
+      -- not the datatype itself
+      qualifier = NE.take (length qnameList - 2) qnameList
```

Comment:
This whole `let` block exists to generate appropriate smart constructor names and I find it a bit ugly, especially the marked line which exists to export the qualified name of the smart constructor _without_ including the name of the data type in the constructor. Is there a better way to do this?